### PR TITLE
graph: Keep types

### PR DESCRIPTION
**Pull request by safchain**

### CODE REVIEW

File: topology/graph/elasticsearch.go [hunk 1]
```diff
@@ -414,8 +395,9 @@ func (b *ElasticSearchBackend) searchNodes(tsq *TimedSearchQuery) (nodes []*Node
 	if out != nil && len(out.Hits.Hits) > 0 {
 		for _, d := range out.Hits.Hits {
 			var node Node
-			if err := b.hitToNode(d.Source, &node); err != nil {
+			if err := common.JSONDecode(bytes.NewReader(*d.Source), &node); err != nil {
```
Comment (safchain) [inline topology/graph/elasticsearch.go]: json.Unmarshal

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by safchain** [comment]
run  skydive-cdd-overview-tests

**Comment by safchain** [comment]
run  skydive-cdd-overview-tests

**Comment by safchain** [comment]
run  skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run  skydive-cdd-overview-tests

**Comment by safchain** [comment]
run  skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run  skydive-functional-tests-backend-elasticsearch

**Comment by safchain** [comment]
run  skydive-functional-tests-backend-orientdb

File: topology/graph/graph.go [hunk 2]
```diff
@@ -469,158 +449,96 @@ func (e *graphElement) MatchMetadata(f ElementMatcher) bool {
...
-func parseTime(i interface{}) (t time.Time, err error) {
-	var ms int64
-	switch i := i.(type) {
-	case int64:
-		ms = i
-	case json.Number:
-		ms, err = i.Int64()
-		if err != nil {
-			return t, err
+func normalizeRawMetadata(raw interface{}) interface{} {
```
Comment (lebauce) [inline topology/graph/graph.go]: nit: rename to normalizeNumbers or normalizeFloats

File: topology/graph/graph.go [hunk 3]
```diff
@@ -606,32 +546,23 @@ func (e *Edge) String() string {
 	return string(b)
 }
 
-// Decode deserialize the current edge
-func (e *Edge) Decode(i interface{}) error {
-	if err := e.graphElement.Decode(i); err != nil {
-		return err
-	}
+// wrapper to avoid unmarshalling infinite loop
+type edgeWrapper Edge
```
Comment (lebauce) [inline topology/graph/graph.go]: Can be moved inside the `Edge.UnmarshalJSON`

File: flow/storage/orientdb/orientdb.go [hunk 4]
```diff
@@ -43,214 +42,188 @@ type Storage struct {
...
 // StoreFlows pushes a set of flows in the database
 func (c *Storage) StoreFlows(flows []*flow.Flow) error {
 	// TODO: use batch of operations
 	for _, flow := range flows {
-		flowDoc, err := c.client.Upsert(flowToDocument(flow), "UUID")
+		fd := flowToDoc(flow)
+		raw, err := json.Marshal(fd)
 		if err != nil {
-			logging.GetLogger().Errorf("Error while pushing flow %s: %s\n", flow.UUID, err.Error())
+			logging.GetLogger().Errorf("Error while pushing flow %s: %s", flow.UUID, err)
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: nit: return fmt.Errorf("Error while pushing flow %s: %s", flow.UUID, err)
and log in the caller

File: flow/storage/orientdb/orientdb.go [hunk 5]
```diff
@@ -43,214 +42,188 @@ type Storage struct {
...
 			return err
 		}
 
-		flowID, ok := flowDoc["@rid"]
-		if !ok {
-			logging.GetLogger().Errorf("No @rid attribute for flow '%s'", flow.UUID)
+		result, err := c.client.Upsert("Flow", json.RawMessage(raw), "UUID", flow.UUID)
+		if err != nil {
+			logging.GetLogger().Errorf("Error while pushing flow %s: %s", flow.UUID, err)
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 6]
```diff
@@ -43,214 +42,188 @@ type Storage struct {
...
+		}
+
+		data := struct {
+			Result []struct {
+				RID string `json:"@rid"`
+			}
+		}{}
+
+		if err := json.Unmarshal(result.Body, &data); err != nil {
+			logging.GetLogger().Errorf("Error while decoding flow %s: %s", flow.UUID, err)
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 7]
```diff
@@ -43,214 +42,188 @@ type Storage struct {
...
+			return err
+		}
+
+		if len(data.Result) == 0 {
+			logging.GetLogger().Errorf("Error while decoding flow %s: no result", flow.UUID)
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 8]
```diff
@@ -43,214 +42,188 @@ type Storage struct {
...
 
 		if flow.LastUpdateMetric != nil {
-			doc := flowMetricToDocument(flow, flow.LastUpdateMetric)
-			doc["Flow"] = flowID
-			if _, err = c.client.CreateDocument(doc); err != nil {
-				logging.GetLogger().Errorf("Error while pushing metric %+v: %s\n", flow.LastUpdateMetric, err.Error())
+			md := metricToDoc(data.Result[0].RID, flow.LastUpdateMetric)
+			raw, err := json.Marshal(md)
+			if err != nil {
+				logging.GetLogger().Errorf("Error while pushing metric %s: %s", flow.UUID, err)
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 9]
```diff
@@ -43,214 +42,188 @@ type Storage struct {
...
+				return err
+			}
+
+			if _, err = c.client.CreateDocument(json.RawMessage(raw)); err != nil {
+				logging.GetLogger().Errorf("Error while pushing metric %+v: %s", flow.LastUpdateMetric, err)
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 10]
```diff
@@ -43,214 +42,188 @@ type Storage struct {
...
 		for _, r := range flow.LastRawPackets {
-			doc := flowRawPacketToDocument(linkType, r)
-			doc["Flow"] = flowID
-			if _, err = c.client.CreateDocument(doc); err != nil {
-				logging.GetLogger().Errorf("Error while pushing raw packet %+v: %s\n", r, err.Error())
+			rd := rawpacketToDoc(data.Result[0].RID, linkType, r)
+			raw, err := json.Marshal(rd)
+			if err != nil {
+				logging.GetLogger().Errorf("Error while pushing raw packet %s: %s", flow.UUID, err)
+				return err
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 11]
```diff
@@ -261,13 +234,22 @@ func (c *Storage) StoreFlows(flows []*flow.Flow) error {
...
 	if err != nil {
 		return nil, err
 	}
 
+	data := struct {
+		Result []*flow.Flow
+	}{}
+
+	if err := json.Unmarshal(result.Body, &data); err != nil {
+		logging.GetLogger().Errorf("Error while decoding flows %s, %s", err, string(result.Body))
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 12]
```diff
@@ -302,24 +284,28 @@ func (c *Storage) SearchRawPackets(fsq filters.SearchQuery, packetFilter *filter
...
-			return nil, err
-		}
-		flowID := doc["Flow"].(string)
+	data := struct {
+		Result []*rawpacketDoc
+	}{}
 
-		if fr, ok := rawpackets[flowID]; ok {
+	if err := json.Unmarshal(result.Body, &data); err != nil {
+		logging.GetLogger().Errorf("Error while decoding flows %s, %s", err, string(result.Body))
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 13]
```diff
@@ -344,19 +330,23 @@ func (c *Storage) SearchMetrics(fsq filters.SearchQuery, metricFilter *filters.F
...
 	if err != nil {
 		return nil, err
 	}
 
+	data := struct {
+		Result []*metricDoc
+	}{}
+
+	if err := json.Unmarshal(result.Body, &data); err != nil {
+		logging.GetLogger().Errorf("Error while decoding flows %s, %s", err, string(result.Body))
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: ditto

File: flow/storage/orientdb/orientdb.go [hunk 14]
```diff
@@ -85,20 +93,27 @@ type metricDoc struct {
...
+		LayersPath:         &f.LayersPath,
+		Application:        &f.Application,
+		Link:               f.Link,
+		Network:            f.Network,
+		Transport:          f.Transport,
+		ICMP:               f.ICMP,
+		Metric:             f.Metric,
+		TCPMetric:          f.TCPMetric,
+		IPMetric:           f.IPMetric,
+		DHCPv4:             f.DHCPv4,
```
Comment (lebauce) [inline flow/storage/orientdb/orientdb.go]: You need to also modify the `Flow` document class in the `New` function

**Comment by lebauce** [comment]
```
diff --git a/analyzer/server.go b/analyzer/server.go
index cf260000..fdb76087 100644
--- a/analyzer/server.go
+++ b/analyzer/server.go
@@ -237,17 +237,9 @@ func NewServerFromConfig() (*Server, error) {
 	// add some global vars
 	uiServer.AddGlobalVar("ui", config.Get("ui"))
 
-	flowMetricKeys, err := (&flow.FlowMetric{}).GetFieldKeys()
-	if err != nil {
-		return nil, err
-	}
-	uiServer.AddGlobalVar("flow-metric-keys", flowMetricKeys)
+	uiServer.AddGlobalVar("flow-metric-keys", (&flow.FlowMetric{}).GetFieldKeys())
 
-	interfaceMetricKeys, err := (&topology.InterfaceMetric{}).GetFieldKeys()
-	if err != nil {
-		return nil, err
-	}
-	uiServer.AddGlobalVar("interface-metric-keys", interfaceMetricKeys)
+	uiServer.AddGlobalVar("interface-metric-keys", (&topology.InterfaceMetric{}).GetFieldKeys())
 
 	uiServer.AddGlobalVar("probes", config.Get("analyzer.topology.probes"))
 
diff --git a/common/types.go b/common/types.go
index 8bc58ac4..fdbbbf4e 100644
--- a/common/types.go
+++ b/common/types.go
@@ -69,7 +69,7 @@ const (
 // Getter describes filter getter fields
 type Getter interface {
 	GetField(field string) (interface{}, error)
-	GetFieldKeys() ([]string, error)
+	GetFieldKeys() []string
 	GetFieldInt64(field string) (int64, error)
 	GetFieldString(field string) (string, error)
 }
@@ -318,7 +318,7 @@ func NewTimeSlice(s, l int64) *TimeSlice {
 type Metric interface {
 	// part of the Getter interface
 	GetFieldInt64(field string) (int64, error)
-	GetFieldKeys() ([]string, error)
+	GetFieldKeys() []string
 
 	Add(m Metric) Metric
 	Sub(m Metric) Metric
@@ -418,9 +418,7 @@ func GetField(obj map[string]interface{}, k string) (interface{}, error) {
 	return obj, nil
 }
 
-func getFieldKeys(obj map[string]interface{}, path string) ([]string, error) {
-	var fields []string
-
+func getFieldKeys(obj map[string]interface{}, path string) (fields []string) {
 	if path != "" {
 		path += "."
 	}
@@ -430,29 +428,19 @@ func getFieldKeys(obj map[string]interface{}, path string) ([]string, error) {
 
 		switch v.(type) {
 		case Getter:
-			keys, err := v.(Getter).GetFieldKeys()
-			if err != nil {
-				return nil, err
-			}
-			for _, subkey := range keys {
+			for _, subkey := range v.(Getter).GetFieldKeys() {
 				fields = append(fields, k+"."+subkey)
 			}
 		case map[string]interface{}:
-			subfields, err := getFieldKeys(v.(map[string]interface{}), path+k)
-			if err != nil {
-				return nil, err
-			}
+			subfields := getFieldKeys(v.(map[string]interface{}), path+k)
 			fields = append(fields, subfields...)
-		case map[interface{}]interface{}:
-			return nil, errors.New("Not supporting map not using string as key")
 		}
 	}
-
-	return fields, nil
+	return
 }
 
 // GetFieldKeys returns all the keys using dot notation
-func GetFieldKeys(obj map[string]interface{}) ([]string, error) {
+func GetFieldKeys(obj map[string]interface{}) []string {
 	return getFieldKeys(obj, "")
 }
 
diff --git a/flow/flow.go b/flow/flow.go
index 02b94d2c..d4145425 100644
--- a/flow/flow.go
+++ b/flow/flow.go
@@ -1388,8 +1388,8 @@ func (f *Flow) GetField(field string) (interface{}, error) {
 }
 
 // GetFieldKeys returns the list of valid field of a Flow
-func (f *Flow) GetFieldKeys() ([]string, error) {
-	return flowFieldKeys, nil
+func (f *Flow) GetFieldKeys() []string {
+	return flowFieldKeys
 }
 
 var flowFieldKeys []string
diff --git a/flow/flow_test.go b/flow/flow_test.go
index 29416204..29c95e5a 100644
--- a/flow/flow_test.go
+++ b/flow/flow_test.go
@@ -1108,13 +1108,8 @@ func TestFlowSimpleSynFin(t *testing.T) {
 }
 
 func TestGetFieldsXXX(t *testing.T) {
-	f := &Flow{}
-
-	fields, err := f.GetFieldKeys()
-	if err != nil {
-		t.Error(err)
-	}
-	for _, k := range fields {
+	var f Flow
+	for _, k := range f.GetFieldKeys() {
 		if strings.HasPrefix(k, "XXX_") {
 			t.Error("XXX_ private field exposed")
 		}
diff --git a/flow/metrics.go b/flow/metrics.go
index a502c086..b016f75b 100644
--- a/flow/metrics.go
+++ b/flow/metrics.go
@@ -176,8 +176,8 @@ func (fm *FlowMetric) Split(cut int64) (common.Metric, common.Metric) {
 }
 
 // GetFieldKeys implements Getter and Metrics interface
-func (fm *FlowMetric) GetFieldKeys() ([]string, error) {
-	return metricsFields, nil
+func (fm *FlowMetric) GetFieldKeys() []string {
+	return metricsFields
 }
 
 var metricsFields []string
diff --git a/gremlin/traversal/flows.go b/gremlin/traversal/flows.go
index 6398ee8b..9de91b0c 100644
--- a/gremlin/traversal/flows.go
+++ b/gremlin/traversal/flows.go
@@ -539,7 +539,7 @@ func (f *FlowTraversalStep) PropertyKeys(ctx traversal.StepContext, keys ...inte
 
 	if len(f.flowset.Flows) > 0 {
 		// all Flow struct are the same, take the first one
-		s, _ = f.flowset.Flows[0].GetFieldKeys()
+		s = f.flowset.Flows[0].GetFieldKeys()
 	}
 
 	return traversal.NewGraphTraversalValue(f.GraphTraversal, s)
diff --git a/gremlin/traversal/metrics.go b/gremlin/traversal/metrics.go
index d98d5fcc..40f4a1f6 100644
--- a/gremlin/traversal/metrics.go
+++ b/gremlin/traversal/metrics.go
@@ -299,7 +299,7 @@ func (m *MetricsTraversalStep) PropertyKeys(ctx traversal.StepContext, keys ...i
 		for _, metrics := range m.metrics {
 			// all Metric struct are the same, take the first one
 			if len(metrics) > 0 {
-				s, _ = metrics[0].GetFieldKeys()
+				s = metrics[0].GetFieldKeys()
 				break
 			}
 		}
diff --git a/topology/graph/graph.go b/topology/graph/graph.go
index 246d455f..b530cad1 100644
--- a/topology/graph/graph.go
+++ b/topology/graph/graph.go
@@ -407,13 +407,9 @@ func (e *graphElement) GetField(name string) (interface{}, error) {
 	}
 }
 
-func (e *graphElement) GetFieldKeys() ([]string, error) {
+func (e *graphElement) GetFieldKeys() []string {
 	keys := []string{"ID", "Host", "CreatedAt", "UpdatedAt", "DeletedAt", "Revision", "Origin"}
-	subkeys, err := common.GetFieldKeys(e.Metadata)
-	if err != nil {
-		return nil, err
-	}
-	return append(keys, subkeys...), nil
+	return append(keys, common.GetFieldKeys(e.Metadata)...)
 }
 
 // GetFieldStringList returns a list a string for the given field name
@@ -519,11 +515,11 @@ func (n *Node) String() string {
 	return string(b)
 }
 
-// wrapper to avoid unmarshal infinite loop
-type nodeWrapper Node
-
 // UnmarshalJSON custom unmarshal function
 func (n *Node) UnmarshalJSON(b []byte) error {
+	// wrapper to avoid unmarshal infinite loop
+	type nodeWrapper Node
+
 	nw := struct {
 		nodeWrapper
 		Metadata map[string]json.RawMessage
diff --git a/topology/graph/traversal/traversal.go b/topology/graph/traversal/traversal.go
index fab7db83..5f0678ca 100644
--- a/topology/graph/traversal/traversal.go
+++ b/topology/graph/traversal/traversal.go
@@ -643,10 +643,7 @@ func (tv *GraphTraversalV) PropertyKeys(ctx StepContext, keys ...interface{}) *G
 
 	seen := make(map[string]bool)
 	for _, n := range tv.nodes {
-		fields, err := n.GetFieldKeys()
-		if err != nil {
-			return NewGraphTraversalValueFromError(err)
-		}
+		fields := n.GetFieldKeys()
 		for _, k := range fields {
 			if _, ok := seen[k]; !ok {
 				s = append(s, k)
diff --git a/topology/metrics.go b/topology/metrics.go
index a2318274..3eff38f1 100644
--- a/topology/metrics.go
+++ b/topology/metrics.go
@@ -308,8 +308,8 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
 }
 
 // GetFieldKeys implements Getter and Metrics interfaces
-func (im *InterfaceMetric) GetFieldKeys() ([]string, error) {
-	return metricsFields, nil
+func (im *InterfaceMetric) GetFieldKeys() []string {
+	return metricsFields
 }
 
 var metricsFields []string
diff --git a/topology/probes/netlink/route.go b/topology/probes/netlink/route.go
index 3913abea..649a4653 100644
--- a/topology/probes/netlink/route.go
+++ b/topology/probes/netlink/route.go
@@ -377,8 +377,8 @@ func (rts *RoutingTables) GetField(field string) (interface{}, error) {
 }
 
 // GetFieldKeys returns the list of valid field of a Flow
-func (rts *RoutingTables) GetFieldKeys() ([]string, error) {
-	return rtFields, nil
+func (rts *RoutingTables) GetFieldKeys() []string {
+	return rtFields
 }
 
 var rtFields []string
diff --git a/topology/probes/socketinfo/connection.go b/topology/probes/socketinfo/connection.go
index 60b9cf99..606479fd 100644
--- a/topology/probes/socketinfo/connection.go
+++ b/topology/probes/socketinfo/connection.go
@@ -108,8 +108,8 @@ func (c *ConnectionInfo) GetField(field string) (interface{}, error) {
 }
 
 // GetFieldKeys returns the list of valid field of a Flow
-func (c *ConnectionInfo) GetFieldKeys() ([]string, error) {
-	return connInfoFieldKeys, nil
+func (c *ConnectionInfo) GetFieldKeys() []string {
+	return connInfoFieldKeys
 }
 
 var connInfoFieldKeys []string
```

File: common/types.go [hunk 15]
```diff
@@ -378,28 +388,37 @@ func GetField(obj map[string]interface{}, k string) (interface{}, error) {
...
 			}
 			return results, nil
-		}
-
-		if obj, ok = i.(map[string]interface{}); !ok {
-			return nil, fmt.Errorf("%s is not a map, but a %+v", component, reflect.TypeOf(obj))
+		case map[string]interface{}:
+			obj = i.(map[string]interface{})
+		default:
+			return nil, fmt.Errorf("%s is not a supported type(%v)", component, reflect.TypeOf(obj))
```
Comment (lebauce) [inline common/types.go]: nit: `type (%+v)`

File: common/types.go [hunk 16]
```diff
@@ -340,7 +350,7 @@ func SetField(obj map[string]interface{}, k string, v interface{}) bool {
 	return true
 }
 
-// DelField detete a value in a tree based on dot key
+// DelField delete a value in a tree based on dot key
```
Comment (lebauce) [inline common/types.go]: s/delete/deletes//

**Comment by masco-enovance** [comment]
tested with nexthop patch, working fine with the routing tables but not able to convert the Neighbors from metadata. getting the following  error ```interface {} is map[string]interface {}, not *netlink.Neighbor```

**Comment by safchain** [comment]
@masco-enovance yes neighbors need to be converted as well, will do it

**Pull request merged**